packages/progress-monitor/src/browser/progress-widget.tsx

```diff
@@ -0,0 +1,132 @@
+/********************************************************************************
+ * Copyright (C) 2018 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject } from 'inversify';
+import { Message } from '@phosphor/messaging';
+import { ReactWidget } from '@theia/core/lib/browser/widgets/react-widget';
+import * as React from 'react';
+import { ProgressMonitorService } from './progress-service';
+import { ProgressReport } from './progress-protocol';
+import { DisposableCollection, Disposable } from '@theia/core/lib/common';
+
+@injectable()
+export class ProgressWidget extends ReactWidget {
+
+    SEARCH_DELAY = 200;
+
+    private progressItems: Map<string, ProgressReport> = new Map();
+
+    protected readonly toDisposeOnSearch = new DisposableCollection();
+
+    constructor(
+        @inject(ProgressMonitorService) protected readonly progressService: ProgressMonitorService
+    ) {
+        super();
+        this.node.classList.add("theia-ProgressWidget");
+        this.update();
+        this.node.onclick = (e) => {
+            e.stopPropagation();
+        };
+        progressService.onContributionAdded((progressReport: ProgressReport) => {
```

Comment:
I would manage the state to `ProgressMonitorService` to separate a model from a widget:
```ts
    progressService.onChanged(() => this.update());
}
```